integration-cli/docker_api_swarm_test.go

```diff
@@ -550,6 +550,32 @@ func (s *DockerSwarmSuite) TestApiSwarmLeaveRemovesContainer(c *check.C) {
 	c.Assert(id, checker.HasPrefix, strings.TrimSpace(id2))
 }
 
+// #23629
+func (s *DockerSwarmSuite) TestApiSwarmLeaveOnPendingJoin(c *check.C) {
+	s.AddDaemon(c, true, true)
+	d2 := s.AddDaemon(c, false, false)
+
+	id, err := d2.Cmd("run", "-d", "busybox", "top")
+	c.Assert(err, checker.IsNil)
+	id = strings.TrimSpace(id)
+
+	go d2.Join("nosuchhost:1234", "", "", false) // will block on pending state
+
+	time.Sleep(1 * time.Second)
```

Comment:
I don't think this would work. We could loop with `100ms` intervals and check when the swarm state changes from `inactive` to `pending`.